Refresh sidebars widgets global after calling wp_set_sidebars_widgets

adamziel opened the pull request:
### Description

Before this PR, batch-saving widgets sometimes moves one of them to "Inactive widgets". The full, step-by-step problem description can be found [here](https://github.com/WordPress/gutenberg/issues/33335#issuecomment-879903958).

Tactically, this PR removes the `retrieve_widget` call that was placed in the API endpoints as a workaround for the following problem:

```
		 * When batch requests are processed, this global is not properly updated by previous
		 * calls, resulting in widgets incorrectly being moved to the wp_inactive_widgets
		 * sidebar.
```

The solution introduced in this PR is correctly refreshing the related global variable in `wp_set_sidebar_widgets`.

### Test plan

* Confirm the attached unit test correctly checks for the regression.
* Confirm all the tests pass.
* If you'd like to do some manual testing – it's hard to trigger the problem from the UI. Instead, switch themes to Twenty twenty one, go to your widgets screen in wp-admin, run the script below in your devtools, and refresh the page – you should have exactly three widgets assigned to the sidebar and zero widgets in "Inactive widgets":

```js
async function giveMeThreeWidgets(){
    
	const sidebars = ['sidebar-1', 'wp_inactive_widgets'];
	async function* getWidgetsIds() {
		for(const sidebar of sidebars) {
				const sidebarWidgets = await window.wp.apiFetch({
					path: `/index.php?rest_route=%2Fwp%2Fv2%2Fsidebars%2F${sidebar}&_locale=user`
				});
				yield* sidebarWidgets.widgets;
		}
	}
	async function toArray(asyncIterator){
		const arr = [];
		for await(const i of asyncIterator) arr.push(i);
		return arr;
	}

	const ids = await toArray(getWidgetsIds());
	return await window.wp.apiFetch( {
		path: "/index.php?rest_route=%2Fbatch%2Fv1&_locale=user", 
		"data": {
			"validation": "require-all-validate", "requests": [
				{ "path": "/wp/v2/widgets","body": {"id_base": "block","instance": { "raw": { "content": "<!-- wp:legacy-widget /-->" } },"sidebar": "sidebar-1"},					"method": "POST"				},
				{ "path": "/wp/v2/widgets", "body": {"id_base": "block", "instance": { "raw": { "content": "<!-- wp:gallery {\"ids\":[10,8,7],\"linkTo\":\"none\"} -->\n<figure class=\"wp-block-gallery columns-3 is-cropped\"><ul class=\"blocks-gallery-grid\"><li class=\"blocks-gallery-item\"><figure><img src=\"http://localhost:8888/wp-content/uploads/2021/07/L-Misc5-1024x576.jpg\" alt=\"\" data-id=\"10\" data-full-url=\"http://localhost:8888/wp-content/uploads/2021/07/L-Misc5.jpg\" data-link=\"http://localhost:8888/?attachment_id=10\" class=\"wp-image-10\"/></figure></li><li class=\"blocks-gallery-item\"><figure><img src=\"http://localhost:8888/wp-content/uploads/2021/07/L-Misc3-1024x576.jpg\" alt=\"\" data-id=\"8\" data-full-url=\"http://localhost:8888/wp-content/uploads/2021/07/L-Misc3.jpg\" data-link=\"http://localhost:8888/?attachment_id=8\" class=\"wp-image-8\"/></figure></li><li class=\"blocks-gallery-item\"><figure><img src=\"http://localhost:8888/wp-content/uploads/2021/07/L-Misc2-1024x640.jpg\" alt=\"\" data-id=\"7\" data-full-url=\"http://localhost:8888/wp-content/uploads/2021/07/L-Misc2.jpg\" data-link=\"http://localhost:8888/?attachment_id=7\" class=\"wp-image-7\"/></figure></li></ul></figure>\n<!-- /wp:gallery -->" }						}, "sidebar": "sidebar-1"					}, "method": "POST"				},
				{ "path": "/wp/v2/widgets","body": {"id_base": "block","instance": { "raw": { "content": "<!-- wp:gallery {\"ids\":[9,7],\"linkTo\":\"none\"} -->\n<figure class=\"wp-block-gallery columns-2 is-cropped\"><ul class=\"blocks-gallery-grid\"><li class=\"blocks-gallery-item\"><figure><img src=\"http://localhost:8888/wp-content/uploads/2021/07/L-Misc4-1024x576.jpg\" alt=\"\" data-id=\"9\" data-full-url=\"http://localhost:8888/wp-content/uploads/2021/07/L-Misc4.jpg\" data-link=\"http://localhost:8888/?attachment_id=9\" class=\"wp-image-9\"/></figure></li><li class=\"blocks-gallery-item\"><figure><img src=\"http://localhost:8888/wp-content/uploads/2021/07/L-Misc2-1024x640.jpg\" alt=\"\" data-id=\"7\" data-full-url=\"http://localhost:8888/wp-content/uploads/2021/07/L-Misc2.jpg\" data-link=\"http://localhost:8888/?attachment_id=7\" class=\"wp-image-7\"/></figure></li></ul></figure>\n<!-- /wp:gallery -->" } },"sidebar": "sidebar-1"},"method": "POST" },
				...ids.map( id => ( { "path": `/wp/v2/widgets/${ id }?force=true`, "method": "DELETE" } ) ),
			]
		},
		"method": "POST",
	} )
}
giveMeThreeWidgets()
```

Trac ticket: https://core.trac.wordpress.org/ticket/53816

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

draganescu commented on src/wp-includes/widgets.php:
```diff
@@ -1042,19 +1042,27 @@ function wp_get_sidebars_widgets( $deprecated = true ) {
- * @param array $sidebars_widgets Sidebar widgets and their settings.
+ * @global array $sidebars_widgets
+ * @param array $new_sidebars_widgets Sidebar widgets and their settings.
+ * @param boolean $call_retrieve_widgets Should call retrieve_widgets?
```
The name of this is too specific: it describes what the code does, but why?

draganescu commented on src/wp-includes/widgets.php:
```diff
@@ -1310,7 +1318,7 @@ function retrieve_widgets( $theme_changed = false ) {
 
 	if ( 'customize' !== $theme_changed ) {
-		wp_set_sidebars_widgets( $sidebars_widgets );
+		wp_set_sidebars_widgets( $sidebars_widgets, false );
```
A comment about why that "false" is snuck in there would be great!

adamziel replied:
@draganescu 👍

adamziel commented:
@draganescu feedback addressed, would you mind re-reviewing please?

draganescu commented:
This looks good now IMO

hellofromtonya commented on src/wp-includes/widgets.php:
```diff
@@ -1042,19 +1042,27 @@ function wp_get_sidebars_widgets( $deprecated = true ) {
  *
  * @global array $_wp_sidebars_widgets
- * @param array $sidebars_widgets Sidebar widgets and their settings.
+ * @global array $sidebars_widgets
+ * @param array $new_sidebars_widgets Sidebar widgets and their settings.
+ * @param boolean $refresh_global_sidebars_widgets Should update the global $sidebars_widgets variable?
```
Formatting for coding standards and param naming to be consistent with other boolean param types:

```suggestion
 * @global array $sidebars_widgets
 *
 * @param array   $new_sidebars_widgets            Sidebar widgets and their settings.
 * @param boolean $refresh_global_sidebars_widgets Optional. Whether to update $sidebars_widgets
 *                                                 global. Default true.
```

hellofromtonya commented on src/wp-includes/widgets.php:
```diff
@@ -1042,19 +1042,27 @@ function wp_get_sidebars_widgets( $deprecated = true ) {
+	// Refresh the $sidebars_widgets global
+	if ( $refresh_global_sidebars_widgets ) {
+		$sidebars_widgets = wp_get_sidebars_widgets();
+		retrieve_widgets( true );
```
```suggestion
		sync_registered_widgets( true );
```

hellofromtonya commented on src/wp-includes/widgets.php:
```diff
@@ -1310,7 +1318,9 @@ function retrieve_widgets( $theme_changed = false ) {
 
 	if ( 'customize' !== $theme_changed ) {
-		wp_set_sidebars_widgets( $sidebars_widgets );
+		// The second parameter is false to avoid recursion: refreshing the global
+		// $sidebars_widgets entails calling retrieve_widgets()
```
Full stops required and multi-line comment format:

```suggestion
		/*
		 * The second parameter is false to avoid recursion: refreshing the global
		 * $sidebars_widgets entails calling retrieve_widgets().
		 */
```

hellofromtonya reviewed (changes requested):
A few coding standard fixes needed. Then I'll do a testing with both the block and classic widgets.

adamziel commented:
@hellofromtonya I committed your suggestions – thank you! There were also some conflicts that I resolved. This should be good to go!

hellofromtonya commented:
## Test Report

**Env:**

- OS: macOS Big Sur 11.5.2
- Browser: Chrome 92.0.4515.159 and Firefox 91.0.2
- Local testing: wp-env (npm/Docker)
- Theme: Twenty Twenty-One
- Plugins: Classic Widgets (activated during that part of the testing)

### Testing Instructions

**Step 1**: Set up the environment:

- Pull the latest `master`
- Build and start

```
npm install
npm run build
npm run env:start
npm run env:install
```

- Log into back-end http://localhost:8889/wp-admin
- Activate Classic Widgets plugin
- Go to Appearance > Widgets

**Step 2**: Interact with widgets and observe: add new ones, reorder, and move some to Inactive. 

Expectation: 
- Add -> adds the new widget. When leaving and returning to the screen, widget is retained
- Reorder -> widgets are reordered and order is retained when leaving screen
- Make inactive -> moves the widget to the Inactive Widgets (no long in the Footer sidebar) and retains widget when leaving screen

**Step 3:** Interact with widgets in Customizer

Expectations:
- Add -> adds new widget and retains it when closing Customizer
- Reorder -> widgets are reordered and order is retained when closing Customizer
- Make inactive (delete) -> widget is removed from Customizer; when closing Customizer, widget appears in the Inactive Widgets area.

**Step 4:** Apply this PR and rebuild

```
npm run grunt patch:https://github.com/WordPress/wordpress-develop/pull/1525
npm run build
```

**Step 5:** Repeat Step 2 and 3 above

Expectations: same expectations as above

### Results

All tests worked as expected ✅ 

No change when applying the patch ✅

hellofromtonya commented on tests/phpunit/tests/rest-api/rest-widgets-controller.php:
```diff
@@ -853,6 +853,72 @@ public function test_create_item_second_instance() {
+			array( 'text-1' )
+		);
+
+		$request = new WP_REST_Request( 'POST', '/wp/v2/widgets' );
```
@adamziel would be helpful to contributors reading this test if there were inline comments to explain the "create" and "delete". For example, here a new widget is being added. On line 900, the first text widget is being deleted.

adamziel replied:
Good idea, done done!

hellofromtonya commented on src/wp-includes/widgets.php:
```diff
@@ -1042,19 +1042,29 @@ function wp_get_sidebars_widgets( $deprecated = true ) {
-	update_option( 'sidebars_widgets', $sidebars_widgets );
+	update_option( 'sidebars_widgets', $new_sidebars_widgets );
+
+	// Refresh the $sidebars_widgets global
```
Nitpick - full-stop for comments (can be fixed during commit).
```suggestion
	// Refresh the $sidebars_widgets global.
```

desrosj commented on src/wp-includes/widgets.php:
```diff
@@ -1327,7 +1337,11 @@ function sync_registered_widgets( $theme_changed = false ) {
-		wp_set_sidebars_widgets( $sidebars_widgets );
+		/*
+		 * The second parameter is false to avoid recursion: refreshing the global
+		 * $sidebars_widgets entails calling retrieve_widgets().
```
This references the old function name, `retrieve_widgets()`.

adamziel replied:
It seems like the name change may not go through so I will leave it that way.

adamziel commented:
@hellofromtonya @desrosj I believe all the feedback on this one is now addressed.

desrosj commented on src/wp-includes/widgets.php:
```diff
@@ -1042,19 +1042,29 @@ function wp_get_sidebars_widgets( $deprecated = true ) {
  */
-function wp_set_sidebars_widgets( $sidebars_widgets ) {
-	global $_wp_sidebars_widgets;
+function wp_set_sidebars_widgets( $new_sidebars_widgets, $refresh_global_sidebars_widgets = true ) {
```
Even though the second parameter is not required, if someone wishes to override `$refresh_global_sidebars_widgets` to `false`, they would need to verify the version is > 5.8.1. I think addressing this in 5.9 and allowing the simplified > 5.9 check (or allowing a plugin to raise the minimum version of WP required to 5.9) would be preferred.

Also, there should be a `@since` annotation in the doc block explaining what changed and **when.

adamziel replied:
> Also, there should be a @since annotation in the doc block explaining what changed and **when.

I just added one! 

I'll resolve this conversation since choosing the version number is now discussed in the the "main" thread.

desrosj reviewed (changes requested):
I'm thinking that since [#53811](https://core.trac.wordpress.org/ticket/53811) has been punted to the 5.9 release, the same should be done for this one. Especially since the `sync_registered_widgets()` function is being discussed further in [Core-53811](https://core.trac.wordpress.org/ticket/53811) and this PR assumes that the function is present in Core.

I also noted inline that I'm apprehensive to make a change to a function's signature in a minor release unless it cannot be avoided.

draganescu commented:
Are we still onboard with this? @desrosj   [#53811](https://core.trac.wordpress.org/ticket/53811) landed in the mean time so this can move forward?

adamziel committed:
Refresh sidebars widgets global after calling wp_set_sidebars_widgets

adamziel committed:
Fix unit tests

adamziel commented:
Thanks for surfacing this PR @draganescu! I just applied all the required changes.

> I'm thinking that since [#53811](https://core.trac.wordpress.org/ticket/53811) has been punted to the 5.9 release, the same should be done for this one. Especially since the sync_registered_widgets() function is being discussed further in [Core-53811](https://core.trac.wordpress.org/ticket/53811) and this PR assumes that the function is present in Core.

@desrosj I reverted to the old, `retrieve_widgets` name that we ended up continuing to use in core.

> I also noted inline that I'm apprehensive to make a change to a function's signature in a minor release unless it cannot be avoided.

Snap, we've missed the 6.0 train with this one. Would you be willing to reconsider including this patch in 6.0.1, or do you want to wait to 6.1?

adamziel committed:
Add the @since annotation

adamziel commented:
The single failing test is unrelated to this PR:

```
Resolved .nvmrc as 14
Found in cache @ /opt/hostedtoolcache/node/14.19.3/x64
/opt/hostedtoolcache/node/14.19.3/x64/bin/npm config get cache
/home/runner/.npm
Error: getCacheEntry failed: Cache service responded with 503
```

hellofromtonya commented:
>Would you be willing to reconsider including this patch in 6.0.1, or do you want to wait to 6.1?

As this is an enhancement and part of a refactoring, it could not go into a minor, but is a candidate for a major such as 6.1. Minors are for bug fixes and security.

adamziel commented:
@hellofromtonya does it need to be added to any list, or do we leave it here and it will get discovered when the time comes?

hellofromtonya commented:
Great question @adamziel (and Hello 👋 ). The [Trac ticket](https://core.trac.wordpress.org/ticket/53816) is already in the 6.1 milestone. And you've refreshed this PR and discussion. As it's in the milestone, it'll also be part of the 6.1 scrub/triage process once the release squad is formed.

Is the updated PR ready for feedback and test report(s)? If yes, some tips: (a) remove the `needs-refresh` keyword and add a comment in Trac that it's ready and (b) drop the PR into the core-test channel for wider testing and test report(s).

costdev commented on tests/phpunit/tests/rest-api/rest-widgets-controller.php:
```diff
@@ -926,6 +926,80 @@ public function test_create_item_second_instance() {
 	}
 
+	/**
+	 * @ticket 53816
```
```suggestion
	 * Tests that running multiple request handlers (create and delete) deletes the intended widget.
	 *
	 * See this comment for more details:
	 * https://github.com/WordPress/gutenberg/issues/33335#issuecomment-879903958
	 *
	 * @ticket 53816
	 * @covers WP_REST_Widgets_Controller::create_item
	 * @covers WP_REST_Widgets_Controller::delete_item
```

1. Copy part of the last inline comment to the docblock for clarity.
2. Add the appropriate `@covers` tags for what is being tested.
3. A test method with multiple assertions should have a 'message' parameter appended to _each_ assertion to add context to a failed assertion.

General example:

```php
$this->assertInstanceOf( $first_post, 'WP_Post', 'the first post was not created' );
$this->assertSame( $first_post, $first_post_cache, 'the cache for the first post was not updated' );
$this->assertInstanceOf( $second_post, 'WP_Post', 'the second post was not created' );
$this->assertSame( $second_post, $second_post_cache, 'the cache for the second post was not updated' );

$expected_posts = array_merge( $initial_posts, array( $first_post, $second_post ) );

$this->assertSameSetsWithIndex( $expected_posts, $final_posts, 'the two new posts are not in the list' ); 
```

costdev commented on tests/phpunit/tests/rest-api/rest-widgets-controller.php:
```diff
@@ -926,6 +926,80 @@ public function test_create_item_second_instance() {
+		$data               = $response->get_data();
+		$data               = $this->remove_links( $data );
+
+		// Confirm that we deleted exactly the widget that we wanted, and
+		// no other one. This tests against a regression in running multiple
+		// request handlers during the same run. See the following comment for more details:
+		// https://github.com/WordPress/gutenberg/issues/33335#issuecomment-879903958
```
The correct format for a multiline comment is:

```php
/*
 * Multiline
 * Comment
 */
```

adamziel replied:
Thank you! I just updated this PR

costdev commented on src/wp-includes/widgets.php:
```diff
@@ -1072,22 +1072,37 @@ function wp_get_sidebar( $id ) {
+ * @global array $sidebars_widgets
+ *
+ * @param array   $new_sidebars_widgets            Sidebar widgets and their settings.
+ * @param boolean $refresh_global_sidebars_widgets Optional. Whether to update $sidebars_widgets
```
```suggestion
 * @param bool  $refresh_global_sidebars_widgets Optional. Whether to update $sidebars_widgets.
```
- Core uses the short type`bool`.
- Added ending period.

costdev reviewed (changes requested):
Thanks for the PR! This review makes some minor suggestions for consistency with Core docblocks and with the test suite.

adamziel committed:
Update src/wp-includes/widgets.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

adamziel committed:
Update tests/phpunit/tests/rest-api/rest-widgets-controller.php

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

adamziel committed:
Use a block comment instead of inline comments

adamziel commented:
Thank you @costdev ! I just applied the changes you suggested

adamziel committed:
Merge branch 'update/refresh_sidebars_widgets_global_after_wp_set_sidebars_widgets' of github.com:adamziel/wordpress-develop into update/refresh_sidebars_widgets_global_after_wp_set_sidebars_widgets

adamziel committed:
Reflow text in the comment block to restart CI checks

costdev commented:
> Thank you @costdev ! I just applied the changes you suggested

@adamziel Thanks! There's also the `message` parameters that need to be added to [the assertions in the test](https://github.com/WordPress/wordpress-develop/blob/d44991f69dced93fc014fc181267e80e7d3532f3/tests/phpunit/tests/rest-api/rest-widgets-controller.php#L999). See [my review comment on this](https://github.com/WordPress/wordpress-develop/pull/1525#discussion_r897052228) (step 3).

> All PHPUnit assertions, as well as all WordPress custom assertions, allow for a `$message` parameter to be passed. This message will be displayed when the assertion fails and can help immensely when debugging a test. This parameter should always be used if more than one assertion is used in a test method. [Handbook Reference](https://make.wordpress.org/core/handbook/testing/automated-testing/writing-phpunit-tests/#using-assertions)

costdev commented on src/wp-includes/widgets.php:
```diff
@@ -1072,22 +1072,37 @@ function wp_get_sidebar( $id ) {
- * @param array $sidebars_widgets Sidebar widgets and their settings.
+ * @global array $sidebars_widgets
+ *
+ * @param array   $new_sidebars_widgets            Sidebar widgets and their settings.
+ * @param bool  $refresh_global_sidebars_widgets Optional. Whether to update $sidebars_widgets.
```
```suggestion
 * @param array   $new_sidebars_widgets            Sidebar widgets and their settings.
 * @param bool  $refresh_global_sidebars_widgets   Optional. Whether to update $sidebars_widgets.
```
- Needed two spaces to align "Optional" with the line above.

costdev commented on src/wp-includes/widgets.php:
```diff
@@ -1072,22 +1072,37 @@ function wp_get_sidebar( $id ) {
- * @param array $sidebars_widgets Sidebar widgets and their settings.
+ * @global array $sidebars_widgets
+ *
+ * @param array   $new_sidebars_widgets            Sidebar widgets and their settings.
+ * @param bool  $refresh_global_sidebars_widgets Optional. Whether to update $sidebars_widgets.
+ *                                                 global. Default true.
```
```suggestion
 * @param array $new_sidebars_widgets            Sidebar widgets and their settings.
 * @param bool  $refresh_global_sidebars_widgets Optional. Whether to update $sidebars_widgets
 *                                               global. Default true.
```
- Needed some realignment for the params after changing `boolean` to `bool`.
- Removed the `.` after `$sidebars_widgets` as this sentence continues with the word "global.".

adamziel committed:
Update src/wp-includes/widgets.php

Needed some realignment for the params after changing boolean to bool.
Removed the . after $sidebars_widgets as this sentence continues with the word "global.".

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

adamziel committed:
Add assertions messages

adamziel committed:
Merge branch 'update/refresh_sidebars_widgets_global_after_wp_set_sidebars_widgets' of github.com:adamziel/wordpress-develop into update/refresh_sidebars_widgets_global_after_wp_set_sidebars_widgets

adamziel commented:
@costdev Ah good note! I just added these messages and committed your suggestion.

audrasjb committed:
Update since mention

costdev reviewed (approved):
The PR looks great to me. I'll try to get this tested again and leave a report on the ticket. :slightly_smiling_face:

mukeshpanchal27 commented on src/wp-includes/widgets.php:
```diff
@@ -1072,22 +1072,37 @@ function wp_get_sidebar( $id ) {
  * Set the sidebar widget option to update sidebars.
  *
  * @since 2.2.0
+ * @since 6.2.0 Added the optional $refresh_global_sidebars_widgets argument.
```
```suggestion
 * @since 6.2.0 Added the optional `$refresh_global_sidebars_widgets` parameter.
```

Update the document format and use parameter instead of argument for consistency.

mukeshpanchal27 commented:
Thanks @adamziel, Left nit-pick feedback.

adamziel committed:
Update src/wp-includes/widgets.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

adamziel commented:
Thanks @mukeshpanchal27!
